File: src/tensorflow/ops/internal/py.nim
import ../../core

{.compile:"../../ops/internal/py/py.cc".}


type EagerPyFunc*[oT:oall] {.header:"../../ops/internal/py/py.h", importcpp:"EagerPyFunc/*'0*/".} = object
  operation*: Operation[oT]
  output*: olist[oT]

proc iieagerPyFunc[oT: oall](scope: Scope,
                  input: olist[oall],
                  token: cstring,
                  Tin: ArraySlice[DType],
                  Tout: ArraySlice[DType],
                  explicitT: type(oT)): EagerPyFunc[oT] {.header:"../../ops/internal/py/py.h", importcpp:"EagerPyFunc(*#, #, tensorflow::string(#), #, #)", constructor.}

proc eagerPyFunc*(scope: Scope,
                  input: olist[oall],
                  token: cstring,
                  Tin: openArray[DType],
                  Tout: openArray[DType],
                  explicitT: type): auto =
  return iieagerPyFunc(scope,
                       input,
                       token,
                       newArraySlice(Tin),
                       newArraySlice(Tout),
                       explicitT)

converter eagerPyFuncToOutList*[oT: oall](op: EagerPyFunc[oT]): olist[oT] {.inline.} = return op.output


type PyFunc*[oT:oall] {.header:"../../ops/internal/py/py.h", importcpp:"PyFunc/*'0*/".} = object
  operation*: Operation[oT]
  output*: olist[oT]

proc iipyFunc[oT: oall](scope: Scope,
             input: olist[oall],
             token: cstring,
             Tin: ArraySlice[DType],
             Tout: ArraySlice[DType],
             explicitT: type(oT)): PyFunc[oT] {.header:"../../ops/internal/py/py.h", importcpp:"PyFunc(*#, #, tensorflow::string(#), #, #)", constructor.}

proc pyFunc*(scope: Scope,
             input: olist[oall],
             token: cstring,
             Tin: openArray[DType],
             Tout: openArray[DType],
             explicitT: type): auto =
  return iipyFunc(scope,
                  input,
                  token,
                  newArraySlice(Tin),
                  newArraySlice(Tout),
                  explicitT)

converter pyFuncToOutList*[oT: oall](op: PyFunc[oT]): olist[oT] {.inline.} = return op.output


type PyFuncStateless*[oT:oall] {.header:"../../ops/internal/py/py.h", importcpp:"PyFuncStateless/*'0*/".} = object
  operation*: Operation[oT]
  output*: olist[oT]

proc iipyFuncStateless[oT: oall](scope: Scope,
                      input: olist[oall],
                      token: cstring,
                      Tin: ArraySlice[DType],
                      Tout: ArraySlice[DType],
                      explicitT: type(oT)): PyFuncStateless[oT] {.header:"../../ops/internal/py/py.h", importcpp:"PyFuncStateless(*#, #, tensorflow::string(#), #, #)", constructor.}

proc pyFuncStateless*(scope: Scope,
                      input: olist[oall],
                      token: cstring,
                      Tin: openArray[DType],
                      Tout: openArray[DType],
                      explicitT: type): auto =
  return iipyFuncStateless(scope,
                           input,
                           token,
                           newArraySlice(Tin),
                           newArraySlice(Tout),
                           explicitT)

converter pyFuncStatelessToOutList*[oT: oall](op: PyFuncStateless[oT]): olist[oT] {.inline.} = return op.output

File: src/tensorflow/ops/internal/py/py.cc
#include "tensorflow/cc/ops/const_op.h"
#include "py.h"

EagerPyFunc::EagerPyFunc(tensorflow::Scope& scope, 
           tensorflow::InputList input, 
           tensorflow::string token, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tin, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tout) {
      if (!scope.ok())
          return;
      auto _input = ::tensorflow::ops::AsNodeOutList(scope, input);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("EagerPyFunc");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "EagerPyFunc")
                                   .Input(_input)
                                   .Attr("token", token)
                                   .Attr("Tin", Tin)
                                   .Attr("Tout", Tout)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      for (tensorflow::int32 i = 0; i < ret->num_outputs(); ++i)
          this->output.push_back(tensorflow::Output(ret, i));
}

PyFunc::PyFunc(tensorflow::Scope& scope, 
           tensorflow::InputList input, 
           tensorflow::string token, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tin, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tout) {
      if (!scope.ok())
          return;
      auto _input = ::tensorflow::ops::AsNodeOutList(scope, input);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("PyFunc");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "PyFunc")
                                   .Input(_input)
                                   .Attr("token", token)
                                   .Attr("Tin", Tin)
                                   .Attr("Tout", Tout)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      for (tensorflow::int32 i = 0; i < ret->num_outputs(); ++i)
          this->output.push_back(tensorflow::Output(ret, i));
}

PyFuncStateless::PyFuncStateless(tensorflow::Scope& scope, 
           tensorflow::InputList input, 
           tensorflow::string token, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tin, 
           tensorflow::gtl::ArraySlice<tensorflow::DataType> Tout) {
      if (!scope.ok())
          return;
      auto _input = ::tensorflow::ops::AsNodeOutList(scope, input);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("PyFuncStateless");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "PyFuncStateless")
                                   .Input(_input)
                                   .Attr("token", token)
                                   .Attr("Tin", Tin)
                                   .Attr("Tout", Tout)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      for (tensorflow::int32 i = 0; i < ret->num_outputs(); ++i)
          this->output.push_back(tensorflow::Output(ret, i));
}

